x/auction/keeper/auctions.go

```diff
@@ -0,0 +1,297 @@
+	macc := k.supplyKeeper.GetModuleAccount(ctx, buyer)
+	if !macc.HasPermission(supply.Minter) { // TODO ideally don't want to import supply
+		return 0, sdk.ErrInternal("module does not have minting permissions")
+	}
+	// store the auction
+	auctionID, err := k.storeNewAuction(ctx, &auction)
+	if err != nil {
+		return 0, err
+	}
+	return auctionID, nil
+}
+
+// StartForwardReverseAuction starts an auction where bidders bid up to a maxBid, then switch to bidding down on price.
+func (k Keeper) StartForwardReverseAuction(ctx sdk.Context, seller string, lot sdk.Coin, maxBid sdk.Coin, otherPerson sdk.AccAddress) (types.ID, sdk.Error) {
+	// create auction
+	auction := types.NewForwardReverseAuction(seller, lot, ctx.BlockTime().Add(types.DefaultMaxAuctionDuration), maxBid, otherPerson)
+
+	// take coins from module account
+	err := k.supplyKeeper.SendCoinsFromModuleToModule(ctx, seller, types.ModuleName, sdk.Coins{lot})
+	if err != nil {
+		return 0, err
+	}
+	// store the auction
+	auctionID, err := k.storeNewAuction(ctx, &auction)
+	if err != nil {
+		return 0, err
+	}
+	return auctionID, nil
+}
+
+// PlaceBid places a bid on any auction.
+func (k Keeper) PlaceBid(ctx sdk.Context, auctionID types.ID, bidder sdk.AccAddress, bid sdk.Coin, lot sdk.Coin) sdk.Error {
+
+	// get auction from store
+	auction, found := k.GetAuction(ctx, auctionID)
+	if !found {
+		return sdk.ErrInternal("auction doesn't exist")
+	}
+
+	// check end time
+	if ctx.BlockTime().After(auction.GetEndTime()) {
+		return sdk.ErrInternal("auction has closed")
+	}
+
+	var err sdk.Error
+	var a types.Auction
+	switch auc := auction.(type) {
+	case types.ForwardAuction:
+		a, err = k.PlaceBidForward(ctx, auc, bidder, bid)
+		if err != nil {
+			return err
+		}
+	case types.ReverseAuction:
+		a, err = k.PlaceBidReverse(ctx, auc, bidder, lot)
+		if err != nil {
+			return err
+		}
+	case types.ForwardReverseAuction:
+		a, err = k.PlaceBidForwardReverse(ctx, auc, bidder, bid, lot)
+		if err != nil {
+			return err
+		}
+	default:
+		panic("unrecognized auction type")
+	}
+
+	// store updated auction
+	k.SetAuction(ctx, a) // TODO maybe move into above funcs
+
+	return nil
+}
+
+func (k Keeper) PlaceBidForward(ctx sdk.Context, a types.ForwardAuction, bidder sdk.AccAddress, bid sdk.Coin) (types.ForwardAuction, sdk.Error) {
+	// Valid New Bid
+	if bid.Denom != a.Bid.Denom {
+		return a, sdk.ErrInternal("bid denom doesn't match auction")
+	}
+	if !a.Bid.IsLT(bid) { // TODO add minimum bid size
+		return a, sdk.ErrInternal("bid not greater than last bid")
+	}
+
+	// Move Coins
+	increment := bid.Sub(a.Bid)
+	bidAmtToReturn := a.Bid
+	if bidder.Equals(a.Bidder) { // catch edge case of someone updating their bid with a low balance
+		bidAmtToReturn = sdk.NewInt64Coin(a.Bid.Denom, 0)
+	}
+	err := k.supplyKeeper.SendCoinsFromAccountToModule(ctx, bidder, types.ModuleName, sdk.NewCoins(bidAmtToReturn.Add(increment)))
```

Comment:
Confirmed zero doesn't cause an error. I would still use swtich/ifelse to avoid these calls when coins is zero so we aren't emitting noisy events for transfering 0 coins.

Reply:
Also, could we just send `bidIncrement` straight to a.Initiator and avoid having to call `SendCoinsFromModuleToModule` on line 142?